### DISCUSSION TITLE
How to set up ZTNET with its own root servers (2)

### DISCUSSION THREAD

**Discussion by gerrymanley** [sinamics/ztnet#262] [Q&A]
Hi,
I have just setup ZTNET on a VPS and its working as it should......very impressed!

Now I would like to set up 2 private root servers but I am having difficulty understanding how to achieve this.

Do I need to set up 2 separate VPS,s with zerotier-one or ZTNET installed and then login to each vps and create a separate planet and then upload each config to my ZTNET controller?
I would really appreciate it if you had an Easy How To for this
Thanks

**Comment by sinamics** [comment]
While I'm not a network or root server expert ( just a developer ), I've prepared some documentation that might guide you through the process. You can find it here: https://ztnet.network/usage/private_root

This guide outlines the steps for setting up private root servers in Ztnet. Please note that it's based on my general knowledge of root servers and configuration, but it should be fairly accurate.

**Reply by gerrymanley** [reply]
Thanks Bernt,
Don't know how I missed that page guide. I'll update here when complete

**Reply by sinamics** [reply]
I just created that page. Let me know how it goes.

**Reply by gerrymanley** [reply]
Thanks Bernt,
I followed your instructions 100% and set up my 2 VPS's as follows:
VPS-1 runs ZTNET and doubles as my Root server nbr 1
VPS-2 runs my Root server nbr 2
I updated 3 servers with the new planet.custom - backed up the original and then CPed planet.custom to planet.
All working fine.
Only one issue which I was aware of prior to updating and that is the Zerotier app no longer works correctly. It shows in app "Online" but does not find ZTnet - I did have it registered to ZTnet prior to creating new Root servers.

Dumb question: Is it possible to have ZTnet connected to both own root servers plus Zerotier's roots.

PS: Great work you done on ZTnet

**Reply by dved123** [reply]
> 
> Dumb question: Is it possible to have ZTnet connected to both own root servers plus Zerotier's roots.
> 

I too am wondering how to achieve this.

**Reply by sinamics** [reply]
yes, thats what moons are for.
https://docs.zerotier.com/roots

**Reply by dved123** [reply]
> yes, thats what moons are for. https://docs.zerotier.com/roots

I guess im wondering what happens if the official planets are unreachable - will the moons created with the world still work? Presumably if the controller for both the custom planet world and the official world are running in private infrastructure that would still work?

**Comment by tinola** [comment]
@gerrymanley, mobile apps don't support custom roots. It's mentioned [here](https://docs.zerotier.com/selfhosting/).

**Reply by gerrymanley** [reply]
Thanks tinola

**Reply by BingMeng** [reply]
Nope, there is the third party solution
https://github.com/kaaass/ZerotierFix

**Reply by PizzaProgram** [reply]
ZeroTierFix is only for Android! It works great, supports planet import from both `file` and `url`.

But the whole **Apple** branch seems to be an impossible task. 
_(Most people can not self-compile apple apps for both iOS, iPad, iMac.)_

Is there anybody that would do that, _(even for a small amount of money)_ until this "dual moon support" is not implemented?
- Should I ask the creator of ZT?
- Or rather should **not** advertise that I'm using a self-hosted planet?

**Comment by n9yty** [comment]
@sinamics It seems this approach is more of a complete severance from the ZeroTier infrastructure than what ZeroTier outlines here:

https://docs.zerotier.com/roots

Here they talk about moons instead of replacing planets. I am wondering what the difference is, and is there any value in setting up the ability in ZTNET to create moons as well?

If you do this, with planets, then you are totally removed from the ZeroTier public network infrastructure, but if you create moons they will be preferred but if not found/available then it falls back on the ZeroTier public infrastructure?  I'm not sure if I have that right. :)

**Comment by sinamics** [comment]
> Here they talk about moons instead of replacing planets. I am wondering what the difference is

Moons are like adding extra servers to the existing ZeroTier network, which can be great for speed and local traffic. They don't replace the main ZeroTier servers (planets), so your network still works with the public ZeroTier network

> if you create moons they will be preferred but if not found/available then it falls back on the ZeroTier public infrastructure? I'm not sure if I have that right. :)

This is my understanding as well.
I'm thinking of adding a feature to ZTNET for creating moons, but I'll need to do some more research to fully understand how it works. Once I get a better grasp of it, I can figure out what's needed to implement this feature.

**Comment by n9yty** [comment]
I believe that when a client is first initiated it calculates the unique node identifier, and that is used for all networks it joins. I would imagine that it does this with the generation of one, contacting the root servers to be sure it isn't already used somehow, and iterating until it gets one.

If we remove the ZeroTier planet / root servers, I would imagine we would want to be sure the node was created before this so that in the future if it ever were to rejoin the public root servers it would not have any conflicts.  If the node id is created as soon as the software is installed then this probably isn't a concern, but I thought I'd at least mention it for your consideration. :). I probably should have done more testing and reading first, but it is a good discussion point.

**Comment by PizzaProgram** [comment]
So, currently:

1. I can **not** use the official servers as "backup", if I'm using a self-hosted root planet. Right?

2. If I add later a 2th, 3th, etc VPS server, using these planet config files, I will have to update the `planet` file on every client one by one?
 _(Or is the server somehow pushing down the new config _(including the new endpoints)_ to all connected clients?)_

3. Do I understand right, that the original planet on ZTnet is actually **a moon**? 
    3.1. If yes, should not the 'ZT Controller' page call it that way, as long as it is not replaced with a _custom planet_?
    3.2. IMHO the "Restore" button should be called that too: [Restore Original Official Moon]     
    ![kép](https://github.com/user-attachments/assets/0d5bb498-9185-4379-ac71-8bc997eab3e9)


4. Should I not list at the **Endpoints** all the 3 opened ports? 
 `Endpoints: 193.201.184.111/9993,193.201.184.111/22333,193.201.184.111/993` ...

**Comment by PizzaProgram** [comment]
# Tested self created root planet + public planet as 2th server!

I've just:
- installed a **second VPS**, 
- copied the new identity back to VPS-1,
- restarted both ZTnet stacks on VPS-1 and VPS-2, 
- downloaded the new planet file,
- installed to a new client, joined the network, authorized on VPS-1

And the end result of `zerotier-cli listpeers` is:

## it seems to joined Both planets! 

But the big question remains for me:

# How can I sync the two server's networks + members?

 - if I log in to the 2th server, it's total empty.
 - or is it mend to be so ?

**Reply by PizzaProgram** [reply]
PS: I've also asked a friend, who has both iOS and macOS devices to test it for me, if he can join with official app to a new network I've created on VPS-2 .
 ...  waiting for his response.

**Reply by PizzaProgram** [reply]
We have tested the **2th VPS** with iOS + iMac. It's working perfect with the dual-planet config.
But _(as I've suspected and wrote about,)_ it has nothing to do with the 1th self-hosted server.

## So basically it's just 2 separate servers in one merged planet file.

The question remains : 
 - How can I sync two _(self-hosted)_ server's networks + members? 

to create a full functioning self-hosted moon, if the 1th (main) server goes offline.
